src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -569,55 +605,161 @@ private BoundPattern BindRecursivePattern(RecursivePatternSyntax node, TypeSymbo
                         BindPattern(subpatternSyntax.Pattern, elementType, isError, diagnostics));
                     patterns.Add(boundSubpattern);
                 }
+
+                return patterns.ToImmutableAndFree();
             }
-            else
+
+            // It is not a tuple type or ITuple. Seek an appropriate Deconstruct method.
+            var inputPlaceholder = new BoundImplicitReceiver(node, declType); // A fake receiver expression to permit us to reuse binding logic
+            BoundExpression deconstruct = MakeDeconstructInvocationExpression(
+                node.Subpatterns.Count, inputPlaceholder, node, diagnostics, outPlaceholders: out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders);
+            deconstructMethod = deconstruct.ExpressionSymbol as MethodSymbol;
+            if (deconstructMethod is null)
             {
```

Comment:
Consider keeping `else` with a single `return patterns.ToImmutableAndFree();` outside of the `if` / `else`.